### PR TITLE
The Big Refactor

### PR DESCRIPTION
**Pull request by gavofyork**

### CODE REVIEW

File: test/BlockChainHelper.cpp [hunk 1]
```diff
@@ -396,29 +410,42 @@ void TestBlock::populateFrom(TestBlock const& _original)
...
-	m_genesisBlock = _genesisBlock;
-	m_lastBlock = m_genesisBlock;
+	m_tempDirBlockchain.reset(new TransientDirectory);
+	ChainParams p = _noProof ? ChainParams(_genesisBlock.getBytes(), _genesisBlock.accountMap()) : ChainParams(genesisInfo(Network::Test), _genesisBlock.getBytes(), _genesisBlock.accountMap());
+	m_blockChain.reset(new BlockChain(p, m_tempDirBlockchain.get()->path(), WithExisting::Kill));
+	if (!m_blockChain->isKnown(BlockHeader::headerHashFromBlock(_genesisBlock.getBytes())))
+	{
+		cdebug << "Not known:" << BlockHeader::headerHashFromBlock(_genesisBlock.getBytes()) << BlockHeader(p.genesisBlock()).hash();
+		cdebug << "Genesis block not known!";
+		throw 0;
```
Comment (CJentzsch) [inline test/BlockChainHelper.cpp]: this looks bad. Why do you throw `0` instead of a real exception?

**Comment by CJentzsch** [comment]
Looked over the test code, looks good.
All tests pass, but filling does not work yet.

File: libethashseal/Ethash.cpp [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+{
+	const unsigned c_expDiffPeriod = 100000;
+
+	if (!_bi.number())
+		throw GenesisBlockCannotBeCalculated();
+	auto minimumDifficulty = chainParams().u256Param("minimumDifficulty");
+	auto difficultyBoundDivisor = chainParams().u256Param("difficultyBoundDivisor");
+	auto durationLimit = chainParams().u256Param("durationLimit");
+	u256 o = max<u256>(minimumDifficulty, _bi.timestamp() >= _parent.timestamp() + durationLimit ? _parent.difficulty() - (_parent.difficulty() / difficultyBoundDivisor) : (_parent.difficulty() + (_parent.difficulty() / difficultyBoundDivisor)));
+//	if (c_network == Network::Olympic)
```
Comment (LefterisJP) [inline libethashseal/Ethash.cpp]: Can this commented out code be removed finally?

File: libethashseal/EthashProofOfWork.cpp [hunk 3]
```diff
@@ -14,45 +14,27 @@
 	You should have received a copy of the GNU General Public License
 	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
 */
-/** @file EthashSealEngine.h
+/** @file Ethash.cpp
```
Comment (LefterisJP) [inline libethashseal/EthashProofOfWork.cpp]: Rename to EthashProofOfWork.cpp

File: libethashseal/EthashProofOfWork.h [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+	cpp-ethereum is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/** @file EthashSealEngine.h
```
Comment (LefterisJP) [inline libethashseal/EthashProofOfWork.h]: Also here rename to EthashProofOfWork.h

File: libethcore/BlockHeader.cpp [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+		m_timestamp = _header[field = 11].toInt<u256>();
+		m_extraData = _header[field = 12].toBytes();
+		m_seal.clear();
+		for (unsigned i = 13; i < _header.itemCount(); ++i)
+			m_seal.push_back(_header[i].data().toBytes());
+	}
+	catch (Exception const& _e)
+	{
+		_e << errinfo_name("invalid block header format") << BadFieldError(field, toHex(_header[field].data().toBytes()));
+		throw;
```
Comment (LefterisJP) [inline libethcore/BlockHeader.cpp:140]: maybe throw a real exception here?

File: libethcore/BlockHeader.h [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+	// Explicit conversion operator
+	operator StoreType() const { return m_flags; }
+	operator std::string() const { return toString(); }
+	bool operator[](Type flag) const { return test(flag); }
+
+	std::string toString() const
+	{
+		std::string str(size(), '0');
+		for (size_t x = 0; x < size(); ++x)
+			str[size()-x-1] = (m_flags & (1<<x) ? '1' : '0');
```
Comment (LefterisJP) [inline libethcore/BlockHeader.h]: spacing between operators

File: libethcore/BlockHeader.h [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+enum Strictness
+{
+	CheckEverything,
+	JustSeal,
+	QuickNonce,
+	IgnoreSeal,
+	CheckNothingNew
+};
+
+/*enum Check
```
Comment (LefterisJP) [inline libethcore/BlockHeader.h:54]: remove commented out code

File: libethcore/ChainOperationParams.cpp [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+	cpp-ethereum is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/** @file ChainOperationParams.h
```
Comment (LefterisJP) [inline libethcore/ChainOperationParams.cpp]: .h -> .cpp

File: libethcore/ChainOperationParams.cpp [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+#include "ChainOperationParams.h"
+#include <libdevcore/Log.h>
+#include <libdevcore/CommonData.h>
+using namespace std;
+using namespace dev;
+using namespace eth;
+
+ChainOperationParams::ChainOperationParams()/*:
+	accountStartNonce(Invalid256)*/
```
Comment (LefterisJP) [inline libethcore/ChainOperationParams.cpp]: remove commented out code

File: libethcore/ChainOperationParams.h [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+	/// Chain params.
+	u256 blockReward = 0;
+	u256 maximumExtraDataSize = 1024;
+	u256 accountStartNonce = 0;
+
+	std::unordered_map<std::string, std::string> otherParams;
+	u256 u256Param(std::string const& _name) const;
+	// Ethash specific:
+	//u256 minGasLimit;
```
Comment (LefterisJP) [inline libethcore/ChainOperationParams.h]: remove commented out code?

File: libethcore/SealEngine.h [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+	virtual bytes sealRLP() const { return bytes(); }
+	virtual StringHashMap jsInfo(BlockHeader const&) const { return StringHashMap(); }
+
+	/// Don't forget to call Super::verify when subclassing & overriding.
+	virtual void verify(Strictness _s, BlockHeader const& _bi, BlockHeader const& _parent = BlockHeader(), bytesConstRef _block = bytesConstRef()) const;
+	/// Don't forget to call Super::populateFromParent when subclassing & overriding.
+	virtual void populateFromParent(BlockHeader& _bi, BlockHeader const& _parent) const;
+
+	bytes option(std::string const& _name) const { Guard l(x_options); return m_options.count(_name) ? m_options.at(_name) : bytes(); }
+	bool setOption(std::string const& _name, bytes const& _value) { Guard l(x_options); try { if (onOptionChanging(_name, _value)) { m_options[_name] = _value; return true; } } catch (...) {} return false; }
```
Comment (LefterisJP) [inline libethcore/SealEngine.h:60]: perhaps split into multiple lines for readability?

File: libethereum/Account.cpp [hunk 12]
```diff
@@ -45,8 +45,15 @@ AccountMap dev::eth::jsonToAccountMap(std::string const& _json, AccountMaskMap*
 
 	js::mValue val;
 	json_spirit::read_string(_json, val);
-
-	for (auto account: val.get_obj().count("alloc") ? val.get_obj()["alloc"].get_obj() : val.get_obj())
+	js::mObject o = val.get_obj();
+/*	if (o.count("genesis"))
```
Comment (LefterisJP) [inline libethereum/Account.cpp]: commented out code needed here?

File: libethereum/BlockChain.cpp [hunk 13]
```diff
@@ -361,6 +387,23 @@ void BlockChain::rebuild(std::string const& _path, std::function<void(unsigned,
...
+{
+	stringstream ss;
+
+	ss << m_lastBlockHash << endl;
+	ldb::Iterator* i = m_extrasDB->NewIterator(m_readOptions);
+	for (i->SeekToFirst(); i->Valid(); i->Next())
+	{
+		ss << toHex(bytesConstRef(i->key())) << "/" << toHex(bytesConstRef(i->value())) << endl;
+	}
+/*	for (auto h = m_lastBlockHash; h; h = info(h).parentHash())
```
Comment (LefterisJP) [inline libethereum/BlockChain.cpp]: remove commented out code?

File: libethereum/MiningClient.cpp [hunk 14]
```diff
@@ -14,31 +14,9 @@
 	You should have received a copy of the GNU General Public License
 	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
 */
-/** @file Params.h
+/** @file EthashClient.cpp
```
Comment (LefterisJP) [inline libethereum/MiningClient.cpp]: EthashClient.cpp - >MiningClient.cpp

File: test/libethereum/BlockChainInsert.cpp [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+
+	cpp-ethereum is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/** @file blockchain.cpp
```
Comment (LefterisJP) [inline test/libethereum/BlockChainInsert.cpp]: blockchain.cpp -> BlockChainInsert.cpp

File: libethcore/BlockHeader.cpp (hunk 5 above)
Comment (gavofyork) [inline libethcore/BlockHeader.cpp:140]: `throw;` when found in a `catch` block simply rethrows the same exception.

Comment (LefterisJP) [inline libethcore/BlockHeader.cpp:140]: totally forgot about that

**Pull request merged**